src/daft-local-execution/src/sinks/outer_hash_join_probe.rs

```diff
@@ -0,0 +1,417 @@
+                            build_side_growable.extend(
+                                build_side_table_idx as usize,
+                                build_row_idx as usize,
+                                1,
+                            );
+                            probe_side_growable.extend(probe_side_table_idx, probe_row_idx, 1);
+                        }
+                    } else {
+                        // if there's no match, we should still emit the probe side and fill the build side with nulls
+                        build_side_growable.add_nulls(1);
+                        probe_side_growable.extend(probe_side_table_idx, probe_row_idx, 1);
+                    }
+                }
+            }
+        }
+        let build_side_table = build_side_growable.build()?;
+        let probe_side_table = probe_side_growable.build()?;
+
+        let join_table = probe_side_table.get_columns(&self.common_join_keys)?;
+        let left = build_side_table.get_columns(&self.left_non_join_columns)?;
+        let right = probe_side_table.get_columns(&self.right_non_join_columns)?;
+        let final_table = join_table.union(&left)?.union(&right)?;
+        Ok(Arc::new(MicroPartition::new_loaded(
+            final_table.schema.clone(),
+            Arc::new(vec![final_table]),
+            None,
+        )))
+    }
+
+    fn finalize_outer(
+        &self,
+        mut states: Vec<Box<dyn StreamingSinkState>>,
+    ) -> DaftResult<Option<Arc<MicroPartition>>> {
+        let states = states
+            .iter_mut()
+            .map(|s| {
+                s.as_any_mut()
+                    .downcast_mut::<OuterHashJoinProbeState>()
+                    .expect("OuterHashJoinProbeSink state should be OuterHashJoinProbeState")
+            })
+            .collect::<Vec<_>>();
+        let tables = states
+            .first()
+            .expect("at least one state should be present")
+            .get_probeable_and_tables()
+            .1;
+
+        let merged_bitmap = {
+            let bitmaps = states
+                .into_iter()
+                .map(|s| {
+                    if let OuterHashJoinProbeState::ReadyToProbe(_, _, bitmap) = s {
+                        bitmap
+                            .take()
+                            .expect("bitmap should be present in outer join")
+                    } else {
+                        panic!("OuterHashJoinProbeState should be in ReadyToProbe state")
+                    }
+                })
+                .collect::<Vec<_>>();
+            bitmaps.into_iter().fold(None, |acc, x| match acc {
+                None => Some(x),
+                Some(acc) => Some(acc.or(&x)),
+            })
+        }
+        .expect("at least one bitmap should be present");
+
+        let mut build_side_growable =
+            GrowableTable::new(&tables.iter().collect::<Vec<_>>(), true, 20)?;
+
+        for (table_idx, row_idx) in merged_bitmap.get_unused_indices() {
```

Comment:
This can be much more performant using a `BitmapIter` https://github.com/Eventual-Inc/Daft/blob/b1ea3b9749e01512f48dfd45f9899a329fc9799f/src/arrow2/src/bitmap/utils/iterator.rs#L8

Which will compress the adjacent valid bits so we can reduce the calls to extend.

Reply:
You can also convert the bitmap into a BooleanArray and use `mask_filter` `https://github.com/Eventual-Inc/Daft/blob/b1ea3b9749e01512f48dfd45f9899a329fc9799f/src/daft-table/src/lib.rs#L321` instead of iterating

Reply:
I thought the BitmapIter returns an iterator over the individual bits? I just checked and there is also `SlicesIterator`: https://github.com/Eventual-Inc/Daft/blob/b1ea3b9749e01512f48dfd45f9899a329fc9799f/src/arrow2/src/bitmap/utils/slice_iterator.rs which is a `Iterator over a bitmap that returns slices of set regions`, did you mean this one?

Reply:
Whoops, yup thats the one

Reply:
Actually, I prefer the BooleanArray as a `mask_filter` method more, it's a lot cleaner. Went with that in the latest commit.